ledger/src/blockstore.rs

```diff
@@ -3229,14 +3202,24 @@ impl Blockstore {
             .collect()
     }
 
-    pub fn get_entries_in_data_block(
+    /// Fetch the entries corresponding to all of the shred indices in `completed_ranges`
+    /// This function takes advantage of the fact that `completed_ranges` are both
+    /// contiguous and in sorted order. To clarify, suppose completed_ranges is as follows:
+    ///   completed_ranges = [..., (s_i, e_i), (s_i+1, e_i+1), ...]
+    /// Then, the following statements are true:
+    ///   s_i < e_i < s_i+1 < e_i+1
+    ///   e_i == s_i+1 + 1
+    fn get_slot_entries_in_block(
         &self,
         slot: Slot,
-        start_index: u32,
-        end_index: u32,
+        completed_ranges: CompletedRanges,
         slot_meta: Option<&SlotMeta>,
     ) -> Result<Vec<Entry>> {
-        let keys: Vec<(Slot, u64)> = (start_index..=end_index)
+        assert!(!completed_ranges.is_empty());
```

Comment:
Adding this `assert`/`panic` here is not ideal; especially after recent blockstore panics on stable branch.

This function is adding an unnecessary presumption that the caller will never send an empty `completed_ranges`. But instead it could just return an empty vector `Ok(vec![])` and let the caller deal with the empty vector (or maybe an explicit error if you don't want to return an empty vector).

The problem here is that even if today's code is upholding that presumption at all the call-sites, someone unknowingly may change the code preceding the call-site in a way that fails the presumption in a very tricky, rare and subtle way and we hit that rare scenario on mainnet.

There are good instances that we need to enforce an invariant in the code with an explicit `assert`, but this is not one of them. You also need to read the implementation to realize the function makes this presumption which is not good.

The `unwrap`s below also could be written in a `panic`-free style.

Reply:
That's fair - I'm open to updating the assert.

> The unwraps below also could be written in a panic-free style.

Regardless of what the exact check is, we explicitly know that `completed_ranges.is_empty() == false` if we get to this point. With that known with certainty, I felt that the `.unwrap()` is fine. What are you suggesting instead; something like this ?
```rust
let (all_ranges_start_index, _) =  *completed_ranges.first().ok().map_err(|e| ...)?;
```

Reply:
Just do
```rust
if let Some(start_index) = completed_ranges.first().copied() else {
   retrun Ok(vec![]);
}
```
Most likely the duplicate check will be inlined and optimized out by the compiler anyways, and even if not, I am sure it won't change tps.

Reply:
Thanks for the response Behzad. I haven't forgotten about this comment; rather, I think something downstream may panic if we return an empty vec here. That is obviously something else that should be hardened, so just want to make sure I understand and can cover that case before adjusting this one

Reply:
Addressed in https://github.com/anza-xyz/agave/pull/1258